File: office-worker/src/handlers/zip-edit.js
// Shared engine behind docx.edit and pptx.edit: both formats are OOXML zip
// archives, so one declarative string/regex-replace-on-named-member engine
// covers both. No script
// execution — ops are data, applied by the fixed src/scripts/zip_edit.py.
import { join } from 'node:path';
import {
  HandlerInputError,
  assertPathWithinJobDir,
  readJobFile,
  runFixedPythonScript,
  writeJobFile,
} from './run-script.js';

function validateOps(ops) {
  if (!Array.isArray(ops) || ops.length === 0) {
    throw new HandlerInputError('params.ops must be a non-empty array');
  }
  for (const op of ops) {
    if (!op || typeof op.member !== 'string' || !op.member) {
      throw new HandlerInputError('every op requires a string "member"');
    }
    if (typeof op.find !== 'string' || typeof op.replace !== 'string') {
      throw new HandlerInputError(`op on member "${op.member}" requires string "find" and "replace"`);
    }
  }
}

export function createZipEditHandler({ mimeType, defaultOutputFilename }) {
  return async function zipEditHandler({ jobDir, files, params, spawn }) {
    const input = files.find((f) => f.fieldname === 'file');
    if (!input) throw new HandlerInputError('a "file" part is required');
    validateOps(params.ops);

    const outputFilename = params.output_filename || defaultOutputFilename;
    assertPathWithinJobDir(jobDir, outputFilename);
    const inputPath = await writeJobFile(jobDir, 'input.zip', input.buffer);
    const opsPath = join(jobDir, 'ops.json');
    await writeJobFile(jobDir, 'ops.json', JSON.stringify(params.ops));
    const outputPath = join(jobDir, 'output.zip');

    await runFixedPythonScript({
      jobDir,
      spawn,
      script: 'zip_edit.py',
      args: [inputPath, opsPath, outputPath],
    });

    const buffer = await readJobFile(jobDir, 'output.zip', 'zip edit did not produce an output archive');
    return {
      outputs: [{ name: 'file', filename: outputFilename, contentType: mimeType, buffer }],
      result: { filename: outputFilename, size: buffer.length },
    };
  };
}

File: office-worker/src/scripts/zip_edit.py
#!/usr/bin/env python3
"""Fixed helper (NOT model-authored) applying declarative string/regex
replace ops to named members of a zip archive. Shared by docx.edit and
pptx.edit — both formats are OOXML zip archives, so one engine covers both
. Zero new node deps: python's stdlib zipfile covers zip
read/write (per the office-doc-tools brief's zero-new-dep steer).

argv: <input.zip> <ops.json> <output.zip>

ops.json is a JSON array of {"member": str, "find": str, "replace": str,
"regex": bool (optional)}. Ops targeting the same member apply in order.
"""
import json
import re
import sys
import zipfile


def apply_ops(text, ops):
    for op in ops:
        find = op["find"]
        replace = op["replace"]
        if op.get("regex"):
            text = re.sub(find, replace, text, flags=re.DOTALL)
        else:
            text = text.replace(find, replace)
    return text


def main():
    input_path, ops_path, output_path = sys.argv[1], sys.argv[2], sys.argv[3]
    with open(ops_path, "r", encoding="utf-8") as f:
        ops = json.load(f)

    by_member = {}
    for op in ops:
        by_member.setdefault(op["member"], []).append(op)

    with zipfile.ZipFile(input_path, "r") as zin:
        names = set(zin.namelist())
        missing = [m for m in by_member if m not in names]
        if missing:
            raise SystemExit(f"member(s) not found in archive: {', '.join(missing)}")

        with zipfile.ZipFile(output_path, "w", zipfile.ZIP_DEFLATED) as zout:
            for info in zin.infolist():
                data = zin.read(info.filename)
                if info.filename in by_member:
                    text = apply_ops(data.decode("utf-8"), by_member[info.filename])
                    data = text.encode("utf-8")
                zout.writestr(info, data)

    # Sanity-validate: the repacked zip must be readable and structurally
    # intact before the handler ships it back as an output artifact.
    with zipfile.ZipFile(output_path, "r") as check:
        bad_member = check.testzip()
        if bad_member is not None:
            raise SystemExit(f"repacked zip failed integrity check at member {bad_member}")


if __name__ == "__main__":
    main()
